Lib/test/test_gc.py

```diff
@@ -959,6 +964,69 @@ def getstats():
 
         gc.enable()
 
+    @unittest.skipIf(ContainerNoGC is None,
+                     'requires ContainerNoGC extension type')
+    def test_trash_weakref_clear(self):
+        # Test that trash weakrefs are properly cleared (bpo-38006).
+        #
+        # Structure we are creating:
+        #
+        #   Z <- Y <- A--+--> WZ -> C
+        #             ^  |
+        #             +--+
+        # where:
+        #   WZ is a weakref to Z with callback C
+        #   Y doesn't implement tp_traverse
+        #   A contains a reference to itself, Y and WZ
+        #
+        # A, Y, Z, WZ are all trash.  The GC doesn't know that Z is trash
+        # because Y does not implement tp_traverse.  To show the bug, WZ needs
+        # to live long enough so that Z is deallocated before it.  Then, if
+        # gcmodule is buggy, when Z is being deallocated, C will run.
+        #
+        # To ensure WZ lives long enough, we put it in a second reference
+        # cycle.  That trick only works due to the ordering of the GC prev/next
+        # linked lists.  So, this test is a bit fragile.
+        #
+        # The bug reported in bpo-38006 is caused because the GC did not
+        # clear WZ before starting the process of calling tp_clear on the
+        # trash.  Normally, handle_weakrefs() would find the weakref via Z and
+        # clear it.  However, since the GC cannot find Z, WR is not cleared and
+        # it can execute during delete_garbage().  That can lead to disaster
+        # since the callback might tinker with objects that have already had
+        # tp_clear called on them (leaving them in possibly invalid states).
+
+        wr_callback_was_run = False
+        def callback(wr):
```

Comment:
We can use a `unittest.mock` object here to simplify this logic:

```diff
diff --git a/Lib/test/test_gc.py b/Lib/test/test_gc.py
index 803191fd69..f37f1e62d1 100644
--- a/Lib/test/test_gc.py
+++ b/Lib/test/test_gc.py
@@ -12,6 +12,7 @@ import textwrap
 import threading
 import time
 import weakref
+import unittest.mock
 
 try:
     from _testcapi import with_tp_del
@@ -996,13 +997,7 @@ class GCTests(unittest.TestCase):
         # since the callback might tinker with objects that have already had
         # tp_clear called on them (leaving them in possibly invalid states).
 
-        wr_callback_was_run = False
-        def callback(wr):
-            # We cannot allow this callback to run.  It could access objects
-            # with invalid states due to tp_clear already being run.  Or, it
-            # could resurrecte objects that were previously considered trash.
-            nonlocal wr_callback_was_run
-            wr_callback_was_run = True
+        callback = unittest.mock.Mock()
 
         class A:
             __slots__ = ['a', 'y', 'wz']
@@ -1024,7 +1019,7 @@ class GCTests(unittest.TestCase):
         del a, wr_cycle
         gc.collect()
         # if the callback runs, we have a bug
-        self.assertFalse(wr_callback_was_run)
+        callback.assert_not_called()
 
 
 class GCCallbackTests(unittest.TestCase):
```